esphome/components/mitsubishi/mitsubishi.cpp

```diff
@@ -44,39 +49,52 @@ const uint8_t MITSUBISHI_BYTE04 = 0X00;
 const uint8_t MITSUBISHI_BYTE13 = 0X00;
 const uint8_t MITSUBISHI_BYTE16 = 0X00;
 
-// Custom Fan Modes
-const char *CLIMATE_FAN_MEDIUM_LOW = "medium_low";
-
-// CLIMATE_FAN_MEDIUM_HIGH = 1;
-
 climate::ClimateTraits MitsubishiClimate::traits() {
   auto traits = climate::ClimateTraits();
   traits.set_supports_action(false);
   traits.set_visual_min_temperature(MITSUBISHI_TEMP_MIN);
   traits.set_visual_max_temperature(MITSUBISHI_TEMP_MAX);
   traits.set_visual_temperature_step(1.0f);
-  traits.set_supported_modes({climate::CLIMATE_MODE_OFF, climate::CLIMATE_MODE_HEAT_COOL, climate::CLIMATE_MODE_DRY});
+  traits.set_supported_modes({climate::CLIMATE_MODE_OFF});
+
   if (this->supports_cool_)
     traits.add_supported_mode(climate::CLIMATE_MODE_COOL);
   if (this->supports_heat_)
     traits.add_supported_mode(climate::CLIMATE_MODE_HEAT);
 
-  // traits.set_supported_modes({climate::CLIMATE_MODE_OFF, climate::CLIMATE_MODE_HEAT_COOL, climate::CLIMATE_MODE_COOL,
-  //                             climate::CLIMATE_MODE_HEAT, climate::CLIMATE_MODE_DRY});
+  if (this->supported_mode_ != MITSUBISHI_OP_MODE_AH)
+    traits.add_supported_mode(climate::CLIMATE_MODE_COOL);
+  if (this->supported_mode_ >= MITSUBISHI_OP_MODE_AHC) {
+    traits.add_supported_mode(climate::CLIMATE_MODE_HEAT_COOL);
+    traits.add_supported_mode(climate::CLIMATE_MODE_HEAT);
+    traits.add_supported_mode(climate::CLIMATE_MODE_COOL);
+  }
+  if (this->supports_dry_ || this->supported_mode_ >= MITSUBISHI_OP_MODE_ADHC)
+    traits.add_supported_mode(climate::CLIMATE_MODE_DRY);
+  if (this->supports_fan_only_ || this->supported_mode_ >= MITSUBISHI_OP_MODE_ADFHC)
+    traits.add_supported_mode(climate::CLIMATE_MODE_FAN_ONLY);
+
+  // Default to only 3 levels in ESPHome even if most unit supports 4. The 3rd level is not used.
   traits.set_supported_fan_modes(
-      {climate::CLIMATE_FAN_LOW, climate::CLIMATE_FAN_MEDIUM, climate::CLIMATE_FAN_HIGH, climate::CLIMATE_FAN_AUTO});
-  if (this->fan_medium_low_)
-    traits.set_supported_custom_fan_modes({CLIMATE_FAN_MEDIUM_LOW});
+      {climate::CLIMATE_FAN_AUTO, climate::CLIMATE_FAN_LOW, climate::CLIMATE_FAN_MEDIUM, climate::CLIMATE_FAN_HIGH});
+  if (this->fan_mode_ == MITSUBISHI_FAN_Q4L)
+    traits.add_supported_fan_mode(climate::CLIMATE_FAN_QUIET);
+  if (/*this->fan_mode_ == MITSUBISHI_FAN_5L ||*/ this->fan_mode_ == MITSUBISHI_FAN_Q4L)
```

Comment:
Thanks for your help and github guidance. Thought Id try this route for comments, which I think is what you suggested. If not let me know if there's a better way.
I think this check should be for `MITSUBISHI_FAN_4L`, which would then add MIDDLE to 4L. I cannot find a check for 4L anywhere at this time.

Reply:
Ok, I think I've understood your comment this time, sorry ;-)

This change should fix it now. Thanks!